src/lib/tls/tls13/tls_channel_impl_13.cpp

```diff
@@ -209,6 +205,32 @@ size_t Channel_Impl_13::received_data(const uint8_t input[], size_t input_size)
       }
    }
 
+void Channel_Impl_13::handle(const Key_Update& key_update)
+   {
+   // make sure Key_Update appears only at the end of a record; see description above
+   if(m_handshake_layer.has_pending_data())
+      { throw Unexpected_Message("Unexpected additional post-handshake message data found in record"); }
+
+   m_cipher_state->update_read_keys();
+
+   // TODO: introduce some kind of rate limit of key updates, otherwise we
+   //       might be forced into an endless loop of key updates.
```

Comment:
Possibly there should be an informational callback so the application can monitor key updates?

(Not required here just wanted to suggest that as a possibility - then the application can enforce rate limiting by whatever mechanism/policy it prefers)